internal/docstore/dynamodocstore/query.go

```diff
@@ -26,75 +26,126 @@ import (
+		if len(f.FieldPath) == 1 && f.FieldPath[0] == c.partitionKey && f.Op == driver.EqualOp {
+			doQuery = true
+			break
+		}
+	}
+	if doQuery {
+		cb = processFilters(cb, q.Filters, c.partitionKey, c.sortKey)
+		ce, err := cb.Build()
+		if err != nil {
+			return nil, nil, err
+		}
+		out, err := c.db.QueryWithContext(ctx, &dynamodb.QueryInput{
+			TableName:                 &c.table,
+			ExpressionAttributeNames:  ce.Names(),
+			ExpressionAttributeValues: ce.Values(),
+			KeyConditionExpression:    ce.KeyCondition(),
+			FilterExpression:          ce.Filter(),
+			ProjectionExpression:      ce.Projection(),
+			ExclusiveStartKey:         startAfter,
+		})
+		if err != nil {
+			return nil, nil, err
+		}
+		return out.Items, out.LastEvaluatedKey, nil
+	} else {
+		if len(q.Filters) > 0 {
+			cb = cb.WithFilter(filtersToConditionBuilder(q.Filters))
+			cbUsed = true
+		}
+		in := &dynamodb.ScanInput{
+			TableName:         &c.table,
+			ExclusiveStartKey: startAfter,
+		}
+		if cbUsed {
+			ce, err := cb.Build()
+			if err != nil {
+				return nil, nil, err
+			}
+			in.ExpressionAttributeNames = ce.Names()
+			in.ExpressionAttributeValues = ce.Values()
+			in.FilterExpression = ce.Filter()
+			in.ProjectionExpression = ce.Projection()
+		}
+		out, err := c.db.ScanWithContext(ctx, in)
+		if err != nil {
+			return nil, nil, err
+		}
+		return out.Items, out.LastEvaluatedKey, nil
 	}
-	return c.db.QueryWithContext(ctx, &dynamodb.QueryInput{
-		TableName:                 &c.table,
-		ExpressionAttributeNames:  ce.Names(),
-		ExpressionAttributeValues: ce.Values(),
-		KeyConditionExpression:    ce.KeyCondition(),
-		FilterExpression:          ce.Filter(),
-		ProjectionExpression:      ce.Projection(),
-		ExclusiveStartKey:         startAfter,
-	})
 }
 
 func processFilters(cb expression.Builder, fs []driver.Filter, pkey, skey string) expression.Builder {
-	// TODO(shantuo): process the partition key and remove the hard-coded key
-	// condition after we fix the API.
-	kbs := []expression.KeyConditionBuilder{expression.KeyEqual(expression.Key(pkey), expression.Value("query"))}
-	var cbs []expression.ConditionBuilder
-
+	var kbs []expression.KeyConditionBuilder
+	var cfs []driver.Filter
 	for _, f := range fs {
 		if kb, ok := toKeyCondition(f, pkey, skey); ok {
 			kbs = append(kbs, kb)
 			continue
 		}
-		cbs = append(cbs, toFilter(f))
+		cfs = append(cfs, f)
 	}
 	keyBuilder := kbs[0]
 	for i := 1; i < len(kbs); i++ {
 		keyBuilder.And(kbs[i])
 	}
 	cb = cb.WithKeyCondition(keyBuilder)
-	var condBuilder expression.ConditionBuilder
-	if len(cbs) > 0 {
-		condBuilder = cbs[0]
-		for i := 1; i < len(cbs); i++ {
-			condBuilder.And(cbs[i])
-		}
-		cb = cb.WithFilter(condBuilder)
+	if len(cfs) > 0 {
+		cb = cb.WithFilter(filtersToConditionBuilder(cfs))
+	}
+	return cb
+}
+
+func filtersToConditionBuilder(fs []driver.Filter) expression.ConditionBuilder {
+	var cb expression.ConditionBuilder
+	if len(fs) == 0 {
+		return cb
```

Comment:
An empty builder will later on result in a build error on the expression builder. We should make sure the fs is not empty, otherwise we might just let it panic since it is unexported.

Reply:
I agree, panic makes sense here. Done.